odoo/addons/base/models/ir_http.py

```diff
@@ -487,3 +490,39 @@ def _response_by_status(self, status, headers, content):
             return werkzeug.utils.redirect(content, code=301)
         elif status != 200:
             return request.not_found()
+
+    @classmethod
+    def _get_session_store(cls):
+        """
+        Get the session store from the root session store
+        """
+        return http.root.session_store
+
+    @api.autovacuum
+    def _session_vacuum(self):
+        """
+        Remove all session file unused for some time based
+        on the ir.config_parameter base.unused_session_age_limit
+
+        Or default to one week in seconds.
+        """
+        Param = self.env['ir.config_parameter']
+        session_store = self._get_session_store()
+        week_in_seconds = 60*60*24*7
+
+        try:
+            session_age_limit = Param.sudo().get_param(
+                'base.unused_session_age_limit',
+                week_in_seconds
+            )
```

Comment:
How does that work in multi-db?

The session store is global to the server, so making this configurable for each db means the lowest limit will apply - which is entirely unexpected from the point of view of individual admins.

In multi-tenant mode this also means any single admin could create a denial of service (on purpose or by mistake) by setting a very low limit.

Reply:
I was thinking of other ways to make it configurable but IMO we could agree to vacuum 1 week old sessions once a day where 1 week old would be hardcoded. If I'm not wrong the problem is that removing old sessions every 1k request is too frequent your hardware @llacroix, would "hardcoded once a day" be acceptable ?

Reply:
> In multi-tenant mode this also means any single admin could create a denial of service (on purpose or by mistake) by setting a very low limit.

This is an ir.cron, and the gc runs in the cron so it wouldn't be possible to really lock a system out. You could simply have multiple cron workers. And only 1 cron job per database can be executed at a time so it wouldn't really cause more issues than it is currently working. I'm not sure how this could create a denial of service. As the cron cannot prevent other from working... or at least you can increase the amount of cron workers to allow concurrent cron to run. 

But it's also not without saying that with the current method, it will cause denial of service already.  If you have a pretty busy server let say, 1000 requests per sec. You can trigger the session gc on the http worker every seconds. If you have enough sessions that it takes more than 1sec to iterate/fstat all files in the session store. Then the following session you risk to have 2 http workers running a gc while you have only 8 remaining http workers processing 1000req/s. What this means is that on the third sec you're more likely to trigger an other gc than when you had 10 available workers... At some point you'll reach a moment where all your http workers are running the gc and 1 worker will respond to around 1000req/sec alone and start a gc... The issue is mainly that odoo doesn't track session and simply loop  over all files and fstat valid and invalid one. If you create 500,000 new sessions per day, you'll never gc more than 500,000 every day. But after a week you'll be always looping over around 4M files of which you'll delete only 500,000 files.  

And that isn't some theory, it's a case we had with some clients that have very busy server and also are having session stores shared on an nfs drive as the odoo was running in a cluster. We had this weird issue where the server wouldn't be responding randomly... Then I found out in the logs that some requests were pretty slow from times to times... but some would exact same request would be slow or fast even thought it was the same request. Then one day on a very very busy day... All workers started to get locked by the gc and the whole cluster got stuck reading the file store directory.

@Julien00859 once a week, once a minute it doesn't really matter as long as it's not being called from the http worker. Like I explained above, if you have enough files, it may lock the http workers from responding for some time. The main advantage of running it from the ir.http vacuum is that we can also inherit the method to remove it completely and implement a custom session storage. For example if you wanted to move away from the session store folder to something like in database session store. 

This fix would pave the way for per database session store instead of a global one. Eventually the session could be part of a model instead of a file... so a gc would be as fast as `delete from ir_session where expiry_date < now`

Reply:
> This is an ir.cron, and the gc runs in the cron so it wouldn't be possible to really lock a system out

Well, what happens if I set `base.unused_session_age_limit` to 1s and then change the frequency of my vacuum cron to run every minute? How will users of other databases be able to work?

This patch may be fine in a single-tenant setup (and you're free to use it as such), but this the official version is used in many multi-tenant situations, where database admins can't be trusted outside of the boundaries of their own database.


> If you have a pretty busy server let say, 1000 requests per sec. You can trigger the session gc on the http worker every seconds.

The issue with the frequency of the "random GC" is known, and if we're looking for fixing that in stable, I suggest introducing an environment parameter that will turn off the `session_gc()` logic entirely. For high-volume systems, we recommend setting up a system (unix) cron that takes care of the session expiration in an efficient manner instead, e.g. something like:
```
find /<path>/Odoo/sessions -type f -mtime +7 -delete
```


PS: For the record, with even higher volumes you will encounter another class of problem, where the number of session files in the session directory reaches the system limit. At that point the session store should use a different directory layout (e.g. split up the store in 256 subdirectories). We plan to merge a fix for that (in master).

Reply:
@odony  ah okay that make sense put this way...

In my opinion it would make sense to have the sessions stored by database in that case so each session store would be database dependent. Currently it's not exactly feasible since the database is known only after the session is obtained... Having the db as part of the session would make a bit more sense as you'd be able to have sessions/[db]/sessions and each db could be independant from each others in a multi tenant setup and it also means any database could have a custom session store per database if that was really necessary. 

For example the openerp session cookie could be an encrypted session_id + dbname then the dispatch method would be able to get the proper registry for the proper database and find the session from there. At this point it could be even storing the sessions inside the database itself.

Reply:
> In my opinion it would make sense to have the sessions stored by database 

Yes, that's in our roadmap, but also completely out of scope for a PR that targets a [stable branch](https://github.com/odoo/odoo/wiki/Contributing#what-does-stable-mean).

If you want to fix the OP issue in 14.0, the simplest and safest course is to introduce an environment parameter to disable the GC (e.g. ODOO_DISABLE_SESSION_GC). When set, the sysadmin would be responsible for the session GC (e.g. with a system cron as previously suggested). That's something we could accept in 14.0, as there will be virtually no risk for existing deployments, while providing a workaround for high-volume cases.

Then if you want to discuss more advanced options for master, feel free to open another PR for that. We're likely to pursue the in-database session store option on our own anyway. So you may not need to do anything but wait a bit.

Reply:
Hello, would this do? https://github.com/odoo/odoo/pull/73314